debug.go

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpc
+
+import (
+	"html/template"
+	"io"
+	"log"
+	"net/http"
+	"sync"
+
+	"golang.org/x/net/trace"
+)
+
+var (
+	dispatchersLock sync.RWMutex
+	dispatchers     []*Dispatcher
+)
+
+func addDispatcherToDebugPages(disp *Dispatcher) {
+	dispatchersLock.Lock()
+	defer dispatchersLock.Unlock()
+
+	dispatchers = append(dispatchers, disp)
+}
+
+func removeDispatcherFromDebugPages(disp *Dispatcher) {
+	dispatchersLock.Lock()
+	defer dispatchersLock.Unlock()
+
+	filtered := dispatchers[:0]
+	for _, x := range dispatchers {
+		if x != disp {
+			filtered = append(filtered, x)
+		}
+	}
+	dispatchers = filtered
```

Comment:
This potentially leaks memory, since the array that's backing the `dispatchers` slice may still be holding a reference to the last `*Dispatcher`. We should use the [standard delete idiom](https://github.com/golang/go/wiki/SliceTricks) instead.